rust/src/table/convert_to_delta.rs

```diff
@@ -0,0 +1,765 @@
+    storage::{DeltaObjectStore, ObjectStoreRef},
+    table::builder::ensure_table_uri,
+    DeltaTable, DeltaTableConfig, DeltaTableError, DeltaTablePartition, ObjectMeta, ObjectStore,
+    ObjectStoreError, Path, Schema, SchemaDataType, SchemaField, NULL_PARTITION_VALUE_DATA_PATH,
+};
+use arrow::{datatypes::Schema as ArrowSchema, error::ArrowError};
+use futures::{future, TryStreamExt};
+use log::{debug, info};
+use parquet::{
+    arrow::async_reader::{ParquetObjectReader, ParquetRecordBatchStreamBuilder},
+    errors::ParquetError,
+};
+use percent_encoding::percent_decode_str;
+use std::{
+    collections::{HashMap, HashSet},
+    num::TryFromIntError,
+    str::Utf8Error,
+    sync::Arc,
+};
+
+/// Error converting a Parquet table to a Delta table
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// Object store error
+    #[error("Object store error: {0}")]
+    ObjectStore(#[from] ObjectStoreError),
+    /// Arrow error
+    #[error("Arrow error: {0}")]
+    Arrow(#[from] ArrowError),
+    /// Parquet error
+    #[error("Parquet error: {0}")]
+    Parquet(#[from] ParquetError),
+    /// DeltaTable error
+    #[error("DeltaTable error: {0}")]
+    DeltaTable(#[from] DeltaTableError),
+    /// Error percent-decoding as UTF-8
+    #[error("Error percent-decoding as UTF-8: {0}")]
+    PercentDecode(#[from] Utf8Error),
+    /// Error converting usize to i64
+    #[error("Error converting usize to i64: {0}")]
+    TryFromUsize(#[from] TryFromIntError),
+    /// Error reaching the end of an iterator
+    #[error("Error reaching the end of an iterator")]
+    EndOfIterator,
+    /// No parquet file is found in the given path
+    #[error("No parquet file is found in the given path")]
+    ParquetFileNotFound,
+    /// Partition column provided by the user does not exist in the parquet files
+    #[error("Partition column provided by the user does not exist in the parquet files")]
+    PartitionColumnNotExist(HashSet<SchemaField>),
+}
+
+/// Convert a Parquet table to a Delta table in place
+///
+/// # Arguments
+///
+/// * `storage` - A shared reference to a [`DeltaObjectStore`] with "/" pointing at delta table root (i.e. where `_delta_log` is located).
+/// * `partition_schema`
+///     - A vector of [`SchemaField`] describing the data type, nullibility and metadata of partion columns.
+///     - If the schema of a partition column is not provided, schema data type is set to string and the field is set to nullable.
+///     - Errors if any partition column provided by the user does not exist in the parquet files.
+pub async fn convert_to_delta(
+    storage: ObjectStoreRef,
+    partition_schema: Vec<SchemaField>,
+) -> Result<DeltaTable, Error> {
+    if storage.is_delta_table_location().await? {
+        info!("A Delta table already exists in the given object store");
+        let mut table = DeltaTable::new(storage, DeltaTableConfig::default());
+        table.load().await?;
+        return Ok(table);
+    }
```

Comment:
Just thinking out lod here... Should this maybe return an error. From the perspective of converting the goven location, the operation is infeasible. On the other hand, converting delta to delta is in fact a no-op :).

Reply:
Updated to return an error when the location is already a Delta table location. This aligns with `ConvertToDeltaCommand` documentation https://books.japila.pl/delta-lake-internals/commands/convert/ConvertToDeltaCommand/#executing-command`